rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+  TypeSpec is registered for that type when a model is loaded via tf.hub.
+
+### Registry vs Protocol vs Base Class
+
+The current RFC proposes that the API for defining TensorFlow extension types be
+implemented using protocols.  Two alternative APIs that we considered are:
+
+* **Base classes**: Define an abstract class that must be included as a base
+  class for TensorFlow extension types.  We could have a single base class with
+  an abstract `type_spec` property and an abstract `dispatch` method; or we
+  could have separate "mix-in" base classes for each one.  These base classes
+  might or might not have metaclasses.
+
+* **Registration**: Provide functions which can be used to register the
+  `TypeSpec` and the dispatch handler for an extension type.
+
+The following table summarizes some of the pros and cons of each approach:
+
+Feature                                       | Protocol   | Subclass | Registry
+--------------------------------------------- | ---------- | -------- | -------
+Works well with type annotations                    | Y    |    Y     |    N
+Extension type doesn't need to depend on TensorFlow | Y    |    N     |    N
+Extension types can be added non-invasively ①      | N    |    N     |    Y
+Provides a centralized list of all extension types  | N    |    Y ②   |    Y
+
+① As noted above, allowing extension types to be added non-invasively has both
+pros and cons.
+
+② Requires defining a metaclass.
+
+
+### TypeSpec.name
+
+`TensorSpec` is currently the only `TypeSpec` that defines a `name` attribute.
+To my knowledge, the main use for this attribute is to override the default
+argument names for concrete functions when they are called with the flat-
+arguments calling convention.  The question has been raised as to whether all
+`TypeSpecs` should have a `name` attribute.  My initial feeling is that they
+should not, because the core concept for `TypeSpec` is a type specification, and
+type specifications are generally not named.  However, feedback on this question
+is welcome, especially when grounded in concrete use cases.  (Does
+`TensorSpec.name` get used for other things than concrete signature argument
+naming?)
+
+
+### Should user-defined types be supported as inputs/outputs when computing gradients?
+
+User-defined types are not currently supported as the inputs or outputs for
+gradients.  I.e., gradients need to be computed with respect to individual
+component tensors.  We could potentially add support for using extension types
+with gradients directly, though further work might be needed to define the exact
+semantics.
+
+
+## Appendix: Changes from Current (Internal-Only) Design
+
+### Changes to `CompositeTensor`
+
+
+* The CompositeTensor base class is replaced with an `ExtensionType` protocol.
+
+* `CompositeTensor._type_spec` is renamed to `ExtensionType.__tf_type_spec__`,
+  and is changed from an abstractproperty to an abstractmethod.
```

Comment:
Will there be a generic way to get the type spec of any tensor-like object? Not only composite tensors, but also regular tensors.

I'm asking because having such a property, method, or magic function for a protocol would allow doing things like:
```python
specs = tf.nest.map_structure(lambda t: t.get_specs(), some_nested_structure)
```

This can help when doing things such as building function output signatures for `tf.map_fn` and other cases when we want to easily build structured nested specs from data.

Reply:
Yes.  `Tensor` will define `__tf_type_spec__`, and `ExtensionType` is replacing `CompositeTensor`.